### DISCUSSION TITLE
I could need a "smarter" solution import task

### DISCUSSION THREAD

**Discussion by HansMeiser999** [microsoft/powerplatform-build-tools#200] [Ideas]
We are storing the solutions in one .git repository. The solution structure is component based, so we have to follow a specific import order. A commit on a solution folder triggers a publish artifact pipeline which publishes the resulting new artifact.

We are having a release distributing the solutions to an internal test system on a nightly basis. My experience is, that the most reliable and stable result is created when a solution is imported managed and upgraded afterwards. This is also necessary as often components are deleted and recreated differently during development iterations due to changed requirements or simply by mistakes.

Therefore I would like to create a release, that only imports a solution when its version is set higher as found in the target organization. When this is not the case, there should be no point in importing it at all. 

Here comes the problem in. Not all solutions are modified every day. So a release with just import + upgrade tasks will fail due to a solution version violation on a solution that was not modified. If I now configure those tasks to run even if previous tasks have failed the result will show up as partially succeeded after it has finished. That is bad in those cases, where the solution import failed due to a problem with the solution itself. In this case it would also show up as partially succeeded and not as failed. 

Therefore I would like to have an option like "Skip import if solution version is not higher than in target organization" that compares both versions and only executes the task if the newer version enables the solution for an upgrade after the import.

There might also be some other way to achieve this like comparing an artifact against earlier releases and if they were used already as just one idea. I am open for everything that would workaround these problems and just let me import with a following upgrade when necessary. This would also lower deployment duration and resource usage.

So if anyone has any idea how to achieve this it would be greatly appreciated.

**Comment by craig-lund** [comment]
I had the same need, so I wrote my own wrapper in PowerShell that gets the version of the solution in the target and gets the version of the solution in source if is there is any *major* or *minor* version increment perform an upgrade, if there is a *build* or *revision* version increment perform update (this is to support patching) if all versions are the same, skip. 

The current problem with this is ```pac solution import``` does not cause an error in PowerShell. #253 . I am continuing to debug this on my end and hope to get an resolution soon.

**Comment by Kevmscotland** [comment]
Using Microsoft.Xrm.Data.PowerShell, connect to the environment to retrieve the solution by name using Get-CrmRecords filtering on the solution entity and get its version.
Then compare this to the solution/artefact version you have in ADO and set a vso task variable to true/false.
Then on the Power Platform Import Solution step, add a condition based on that variable value.

**Comment by petrochuk** [comment]
We are adding new flag `--skip-lower-version` to `solution import`

![image](https://github.com/microsoft/powerplatform-build-tools/assets/30735471/9f3443f1-e748-40f8-961d-20abbab55680)

**Reply by rajyraman** [reply]
@petrochuk - Will this be in 1.24.4?

**Reply by petrochuk** [reply]
1.25 and above which we plan to release this month
